src/dtgtk/thumbtable.c

```diff
@@ -475,44 +475,97 @@ static gboolean _thumbtable_update_scrollbars(dt_thumbtable_t *table)
+  GList *l = (*th_invalid);
   while(l)
   {
     dt_thumbnail_t *th = (dt_thumbnail_t *)l->data;
-    if(th->y + table->thumb_size <= 0 || th->y > table->view_height
-       || (table->mode == DT_THUMBTABLE_MODE_FILMSTRIP
-           && (th->x + table->thumb_size <= 0 || th->x > table->view_width)))
+    *th_invalid = g_list_remove(*th_invalid, th);
+    table->list = g_list_remove(table->list, th);
+    gtk_container_remove(GTK_CONTAINER(gtk_widget_get_parent(th->w_main)), th->w_main);
+    dt_thumbnail_destroy(th);
+    changed++;
+    l = (*th_invalid);
+  }
+  return changed;
+}
+
+static void _thumb_move_or_create(dt_thumbtable_t *table,
+                                  GList **th_invalid,
+                                  const int imgid,
+                                  const int rowid,
+                                  const int posx,
+                                  const int posy,
+                                  const gboolean top)
+{
+  if(!th_invalid || g_list_length(*th_invalid)==0)
+  {
+    // no widgets available. We need to create one
+    dt_thumbnail_t *thumb = dt_thumbnail_new(table->thumb_size,
+                                             table->thumb_size,
+                                             IMG_TO_FIT,
+                                             imgid,
+                                             rowid,
+                                             table->overlays,
+                                             DT_THUMBNAIL_CONTAINER_LIGHTTABLE,
+                                             table->show_tooltips);
+    if(table->mode == DT_THUMBTABLE_MODE_FILMSTRIP)
     {
-      table->list = g_list_remove_link(table->list, l);
-      gtk_container_remove(GTK_CONTAINER(gtk_widget_get_parent(th->w_main)), th->w_main);
-      dt_thumbnail_destroy(th);
-      g_list_free(l);
-      l = table->list;
-      changed++;
+      thumb->single_click = TRUE;
+      thumb->sel_mode = DT_THUMBNAIL_SEL_MODE_MOD_ONLY;
     }
+    thumb->x = posx;
+    thumb->y = posy;
+    if(top)
+      table->list = g_list_prepend(table->list, thumb);
     else
-    {
-      l = g_list_next(l);
-    }
+      table->list = g_list_append(table->list, thumb);
+
+    // we remember image margins for new thumbs (this limit flickering)
+    dt_thumbnail_t *first = (dt_thumbnail_t *)table->list->data;
+    gtk_widget_set_margin_start(thumb->w_image_box, gtk_widget_get_margin_start(first->w_image_box));
+    gtk_widget_set_margin_top(thumb->w_image_box, gtk_widget_get_margin_top(first->w_image_box));
+    gtk_layout_put(GTK_LAYOUT(table->widget), thumb->w_main, posx, posy);
+  }
+  else
+  {
+    // let's reuse a now unaffected widget
+    dt_thumbnail_t *thumb = (dt_thumbnail_t *)(*th_invalid)->data;
+    thumb->imgid = imgid;
+    thumb->rowid = rowid;
+    thumb->x = posx;
+    thumb->y = posy;
+    dt_thumbnail_reload_infos(thumb);
+    if(thumb->img_surf && cairo_surface_get_reference_count(thumb->img_surf) > 0)
```

Comment:
I'm not going to review all this code because it requires auditing the rest of thumbtable. This is not a new pattern (it occurs a few times in `thumbnail.c` (where it probably belongs; exposing and using those internals elsewhere is not great; publish functions if need be) but could you please explain why this reference count check is needed? Supposedly if somehow the surface has no references it may have been destroyed/cleaned up completely already anyway and if we still have a pointer, this could be dangling and we shouldn't use it to check a phantom object. If we took a reference before storing the pointer, then we can just safely destroy it here.

What I'm saying is `destroy` already decreases and checks outstanding references
https://www.cairographics.org/manual/cairo-cairo-surface-t.html#cairo-surface-destroy

Reply:
this comes from the fact that *sometimes* the surface is in an indeterminate state. certainly because of a race condition... After *lot* of try and guess, we have only found this check to avoid crash... But I admit that this is quite dirty...
My understanding is that when destroying a surface, the reference count is decreased at the far end of the process. Before the surface is in a inderminated state... `cairo_surface_status` doesn't help in this case...

Reply:
And yes, I agree with you and I'll add a specific fct for all this, with some comments...

Reply:
> My understanding is that when destroying a surface, the reference count is decreased at the far end of the process.

If that's the case, how would testing for the reference count help?

I understand these kind of issues can be really aggravating and bugs in gtk _do_ exist.

Reply:
> certainly because of a race condition

All of the accesses to gtk objects are only from the ui thread, correct?

Reply:
I've read again the code (because tbh, I don't remember exactly all the story)
What I would say : sometimes (always ?) when the surface is destroyed by gtk directly for whatever reason, the pointer is not set to NULL *but* not valid either (calling `cairo_surface_destroy` raise an error/crash) and `cairo_surface_status` doesn't help either...
So the only reliable way we have found is to get the reference counting...

Now as you say, this *may* be a gtk bug... don't know...

Reply:
> when the surface is destroyed by gtk directly for whatever reason

What? Gtk (or I guess you mean cairo) goes about randomly destroying surfaces behind our backs? That should cause problems all over the place. But it seems only here...

My worry is that maybe at some point there was a bug where a surface got unintentionally destroyed somewhere without clearing the pointer and the bug was "fixed" by adding checks everywhere. And then eventually the bug was really fixed but the checks stayed and are now still being copied all over the place "just to be safe". This doesn't make sense and really complicates code maintenance and maybe even hides real new bugs somewhere (or makes them much harder to spot).

Every time I'm trying to change things in convoluted code I end up having to clean up all kinds of unnecessary crap before the wood can be seen for the trees again. So this is a real serious worry for me. Althoug of course this isn't at all the worst instance.

I see that these checks were introduced originally as part of your massive (130 commit) rewrite. Could you at all remember if crashes could be triggered reliably before these checks? Could we just try removing them and seeing if that works OK, as expected? And if not, I could help track down and solve the real causes of the problem?

Reply:
tbh, that was a long time ago...
You're right, maybe the best is to remove all that checks and see if we manage to reproduce the crash... I'll try to do that asap...